### ISSUE TITLE
prec argument in MakeNewformModFrm

### ISSUE THREAD

**Issue by edgarcosta** [LMFDB/lmfdb#4354]
Our default argument for `prec` in `MakeNewformModFrm` seems to be too low.

See, for example, https://beta.lmfdb.org/ModularForm/GL2/Q/holomorphic/25/2/e/a/.
where we have `prec = 5` which matches the Sturm bound for the newspace: https://beta.lmfdb.org/ModularForm/GL2/Q/holomorphic/25/2/e/

However, when using the default argument for precision, we do not generate the right form.
As you can see below, `g` does not match https://beta.lmfdb.org/ModularForm/GL2/Q/holomorphic/25/2/e/a/19/2/, but `f` does.

I don't understand why 5 isn't enough.

```
> g := MakeNewformModFrm_25_2_e_a();
> ComplexEmbeddings(g)[1][1];
q + (0.174207000196738539080148396567 - 0.0566032855855701387158634781055*$.1)*q^2 - (0.865190005821791114977854979372 + 1.19083188178546266893673037053*$.1)*q^3 - (1.59088984177142999905214945448 - 1.15584912741866332860038006911*$.1)*q^4 + (0.107665847199744035869707640694 + 2.23347443803298572010538348343*$.1)*q^5 - (1.08143244093539454745731213963 + 6.40936194050502365157210824452*$.1)*q^8 + (0.0393970060185296540580033759634 - 0.951056516295153572116439333374*$.1)*q^9 + (2.21715317582771879321949903909 + 10.1406814663632427297346012327*$.1)*q^11 + O(q^12)
> f := MakeNewformModFrm_25_2_e_a(:prec:=1000);
> ComplexEmbeddings(f)[1][1];
q + (0.174207000196738539080148396567 - 0.0566032855855701387158634781055*$.1)*q^2 - (0.865190005821791114977854979372 + 1.19083188178546266893673037053*$.1)*q^3 - (1.59088984177142999905214945448 - 1.15584912741866332860038006911*$.1)*q^4 + (0.107665847199744035869707640694 + 2.23347443803298572010538348343*$.1)*q^5 - (0.218127152603517425050143962717 + 0.158478652879170686463792643008*$.1)*q^6 + (3.15544362088404722164691426113E-29 - 3.26085561707999236067471813508*$.1)*q^7 - (0.427050983124842272306880251543 - 0.587785252292473129168705954668*$.1)*q^8 + (0.257524158622047079108147338680 - 0.792577863415982885652646690366*$.1)*q^9 + (0.145178135728359697357024214256 + 0.382992641168963372684765218907*$.1)*q^10 + (0.618033988749894848204586834341 + 1.90211303259030714423287866668*$.1)*q^11 + O(q^12)
```

**Comment by edgarcosta** [comment]
@jvoight @AndrewVSutherland, any idea why 5 is not enough?

**Comment by jvoight** [comment]
Drew isolated the issue: the confusion is that Magma's ModFrm computes S_k^new(N,[chi]), the direct sum over a Galois orbit of characters, whereas the Sturm bound applies to S_k^new(N,chi) for any chi in [chi].  

Note this is (or at least should be) resolved by using ModSym instead of ModFrm, and also the ModSym code is often much faster, because it doesn't have to carve out something from a larger space.  Unfortunately, there is currently no function which eats a space of modular symbols ModSym and makes the corresponding space of modular forms ModFrm--only the other way around.  This doesn't make sense, because ModFrm is just a wrapper for ModSym; should we report this to Magma?  

A second approach is to cut out S_k(N,chi) inside S_k(N,[chi]) using diamond operators.  Then we apply the Sturm bound to the subspace S_k(N,chi).  To that end, we need a set of primes p coprime to N generating (ZZ/N*ZZ)^* (better: generating (ZZ/N*ZZ)^* modulo the kernel of chi), and then we demand that
  T_{p^2} = (T_p)^2 - p^(k-1)*chi(p)
for such p.  Recall we have the formal identity
  T_{p^2} = (T_p)^2 - p^(k-1)*<p>
where <p> is the diamond operator.  So the above equalities are
equivalent to insisting that
  <p>(f) = chi(p)*f
for p generating (ZZ/N*ZZ)^*, which I think implies that
  <n>(f) = chi(n)*f
for all n coprime to N, and that's precisely the condition that f has
character chi.  Unfortunately, Magma doesn't compute Hecke operators for these spaces!  

```
> HeckeOperator(S,p^2) - HeckeOperator(S,p)^2 + p^(k-1)*chi(p);

>> HeckeOperator(S,p^2) - HeckeOperator(S,p)^2 + p^(k-1)*chi(p);
                ^
Runtime error in 'HeckeOperator': Hecke operator computation currently only 
supported for spaces with a single character that takes values +/-1.
```

**Comment by AndrewVSutherland** [comment]
Another option would be to keep the current linear algebra approach and just make the precision big enough to ensure the linear system has full rank -- the newform f we want can be written as a Q(f)-linear combination of the echelonized Z-basis that Magma function Basis returns for S_k(N,[chi]), and if we use the coefficients of f to construct a linear system with a unique solution, it must yield the correct form.

This should be easy to implement and will work fine in spaces where the dimension is reasonably small which is most of the spaces where we have forms with exact eigenvalues, and in particular to all the weight 1 spaces.  It might get painful in spaces of large dimension that happen to contain a newform of small dimension, such as https://beta.lmfdb.org/ModularForm/GL2/Q/holomorphic/356/3/n/.

For weight k > 1 we could consider just providing the MakeModSym function and not providing MakeModFrm at all, I think anything one wants to do with a Magma modular form of weight k > 1 can be done using the corresponding irreducible subspace of modular symbols.

**Comment by jvoight** [comment]
Yup, sounds like we ended up in the same place!  

I don't see why we should take away functionality for ModFrm, maybe just add a warning like "MakeModSym will likely be faster for weight k >= 2"?

**Comment by AndrewVSutherland** [comment]
Sure, in fact simply reording the functions in the download file might achieve this (if you cut and paste the code the last thing you see is currently MakeModFrm, which as Edgar noted is likely to make you inclined to use it).

We could add a function MakeQSeries(n) that returns the Q-expansion to as many places as you want by calling MakeModSym for k > 1 and MakeModFrm for k = 1.  My guess is that most of the time people want to create the modular form or space of modular symbols it is so they can extract more coefficientes than we store.

**Comment by edgarcosta** [comment]
as a hotfix, we should do disable the optional argument for prec, use the maximum precision that we can, and then add an assert that the system was uniquely determined.

also, we should move MakeModSym to the bottom.

**Comment by AndrewVSutherland** [comment]
Addressed by #4378
